pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java

```diff
@@ -174,7 +175,14 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
       }
 
       if (loadFSTIndex) {
-        _fstIndex = new LuceneFSTIndexReader(segmentReader.getIndexFor(columnName, ColumnIndexType.FST_INDEX));
+        PinotDataBuffer buffer = segmentReader.getIndexFor(columnName, ColumnIndexType.FST_INDEX);
+
+        if (org.apache.pinot.segment.local.utils.nativefst.FSTHeader
+            .isValidFST(buffer.toDirectByteBuffer(0, (int) buffer.size()))) {
```

Comment:
Can we directly read the first int and compare it with the magic header? That way we can avoid creating a `ByteBuffer` out of the existing buffer. Also in certain extreme case `buffer.size()` could overflow when casting to `int`.